### ISSUE TITLE
strict-boolean-expression false negative

### ISSUE THREAD

**Issue by OliverJAsh** [palantir/tslint#2757]
### Bug Report

- __TSLint version__: 5.2.0
- __TypeScript version__: 2.3.2
- __Running TSLint via__: (pick one) CLI

#### TypeScript code being linted

``` ts
// fp-ts is the npm module, version 0.2.7
import { array } from 'fp-ts';

const hasTweetsAfterRangeEnd = (tweets: {}[]) => (
    array.last(tweets)
        .map(() => true)
        .getOrElse(() => false)
);

export const isLastPage = () => (tweets: {}[]) => {
    return !hasTweetsAfterRangeEnd(tweets);
};
```

with `tslint.json` configuration:

```json
{
    "rules": {
        "no-unused-variable": true,
        "strict-boolean-expressions": true
    }
}
```

and `tsconfig.json`:

``` json
{
    "compilerOptions": {
        "module": "commonjs",
        "target": "es6",
        "strictNullChecks": true
    },
    "files": [
        "./src/publication.ts"
    ]
}
```

#### Actual behavior

``` ts
❯ tslint --project tsconfig.json --type-check ./src/publication.ts

ERROR: src/publication.ts[10, 13]: This type is not allowed in the operand for the '!' operator because it is always truthy. Only booleans are allowed.
```

#### Expected behavior

No errors.

#### Notes

If I remove `no-unused-variable`, the error disappears.

**Comment by OliverJAsh** [comment]
Also, if I make the return type of `hasTweetsAfterRangeEnd` explicit as `boolean`, the error disappears.
